transpiler/transpiler.go

```
@@ -0,0 +1,562 @@
+	}
+	l, ok := vs.Values[0].(*ast.BasicLit)
+	if !ok {
+		return token.ILLEGAL, "", fmt.Errorf("unsupported value: %#v", vs.Values[0])
+	}
+	return l.Kind, l.Value, nil
+}
+
+func isValueConst(vs *ast.ValueSpec) bool {
+	return vs.Names[0].Obj.Kind == ast.Con
+}
+
+// handleValueSpec handles a file level a constant or variable.
+func handleValueSpec(out *output, vs *ast.ValueSpec) error {
+	if len(vs.Names) == 0 {
+		return out.Errorf(vs, "unsupported # of value names: %v", vs.Names)
+	}
+	var decl []string
+	kind, lit, err := guessType(vs)
+	if err != nil {
+		return out.Errorf(vs, "%s", err)
+	}
+	isConst := isValueConst(vs)
+	typ := tokenStr(kind, isConst)
+	if len(typ) == 0 {
+		return out.Errorf(vs, "unsupported literal kind: %s", kind)
+	}
+	// Strictly speaking the C++ version could also define all the variables on
+	// one line but the following is easier to implement.
+	for _, name := range vs.Names {
+		out.Writef(vs, "%s;\n", strings.Join(append(decl, typ, name.Name, "=", lit), " "))
+	}
+	return nil
+}
+
+func tokenStr(kind token.Token, isConst bool) string {
+	switch kind {
+	case token.INT:
+		if isConst {
+			return "const int"
+		}
+		return "int"
+	case token.STRING:
+		if isConst {
+			return "const char * const"
+		}
+		return "const char *"
+	default:
+		return ""
+	}
+}
+
+// nodeType returns the string representation of the type of a node, in C
+// language.
+func nodeType(n ast.Node) (string, error) {
+	if _, ok := n.(*ast.InterfaceType); ok {
+		return "void *", nil
+	}
+	if i, ok := n.(*ast.Ident); ok {
+		return i.Name, nil
+	}
+	if _, ok := n.(*ast.FuncType); ok {
+		return "", fmt.Errorf("unsupported func type")
+	}
+	return "", fmt.Errorf("unsupported %s", reflect.TypeOf(n))
+}
+
+func identName(x ast.Expr) string {
+	switch t := x.(type) {
+	case *ast.SelectorExpr:
+		return identName(t.X) + "." + identName(t.Sel)
+	case *ast.Ident:
+		return t.Name
+	default:
+		return "<N/A>"
```

should we return an error?